subprojects/model-core/src/main/java/org/gradle/api/internal/provider/BiProvider.java

```diff
@@ -41,7 +42,11 @@ public String toString() {
 
     @Override
     public boolean calculatePresence(ValueConsumer consumer) {
-        return left.calculatePresence(consumer) && right.calculatePresence(consumer);
+        if (!left.calculatePresence(consumer) || !right.calculatePresence(consumer)) {
+            return false;
+        }
+        // Purposefully only calculate full value if left & right are both present, to save time
+        return !calculateValue(consumer, false).isMissing();
```

Comment:
Yes. There is no way around it though, allowing `null` to remove the value means that the value must be calculated first. If this is not acceptable in 7.6, this will have to be put in 8.0.

There is no caching implied by `zip` so I am not sure the exact time of calculation breaks anything. It will just recalculate again later.

Reply:
This is indeed a potentially breaking change, but it is hard for me to judge whether this qualifies as something the requires to go into next major. Maybe we should consult some other folks about this.

Regadless of this, as noted in the other comment, we don't need a separate flag for side effects, it could just be
```
return !calculateOwnValue(consumer).isMissing()
```